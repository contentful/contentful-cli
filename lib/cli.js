const yargs = require('yargs')
const { log } = require('./utils/log')
const {
  buildContext,
  getCommand,
  assertContext
} = require('./utils/middlewares')
const { version } = require('../package.json')

yargs
  .usage('\nUsage: contentful <cmd> [args]')
  .commandDir('cmds')
  .middleware([getCommand, buildContext, assertContext])
  .scriptName('')
  .demandCommand(4, 'Please specify a command.')
  .help('h')
  .alias('h', 'help')
  .strict()
  .recommendCommands()
  .option('v', {
    alias: 'version',
    global: false,
    type: 'boolean',
    describe: 'Show current version',
    skipValidation: true
  })
  .version(false)
  .epilog('Copyright 2019 Contentful')
  .fail(function(msg, err, yargs) {
    if (err) throw err
    console.error(yargs.help())
    console.error(msg)
    process.exit(1)
  })
  .parse(process.argv.slice(2), (_, argv, output) => {
    if (argv.version === true && !argv._.length) {
      log(version)
    } else {
      log(output)
    }
  })
