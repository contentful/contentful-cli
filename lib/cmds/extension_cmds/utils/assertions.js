const { kebabCase } = require('lodash')

const { ValidationError } = require('../../../utils/error')

async function assertHasRequiredProperties(extension, required) {
  const missingRequiredFields = required.reduce((missing, field) => {
    return Object.prototype.hasOwnProperty.call(extension, field)
      ? missing
      : [...missing, field]
  }, [])

  if (missingRequiredFields.length) {
    throw new ValidationError(
      `Missing required properties: ${missingRequiredFields
        .map(kebabCase)
        .join(', ')}`
    )
  }
}

async function assertHasOneProperty(extension, list) {
  const filtered = list.filter(prop => {
    return (
      Object.prototype.hasOwnProperty.call(extension, prop) && !extension.prop
    )
  })

  if (filtered.length === list.length || filtered.length === 0) {
    throw new ValidationError(`Must contain exactly one of: ${list.join(', ')}`)
  }
}

async function assertForceOrCorrectVersionProvided(data, targetVersion) {
  if (!data.force && !data.version) {
    throw new ValidationError(
      'Please provide current version or use the --force flag'
    )
  }

  if (data.version && data.version !== targetVersion) {
    throw new ValidationError(
      'Version provided does not match current resource version'
    )
  }
}

module.exports.assertForceOrCorrectVersionProvided =
  assertForceOrCorrectVersionProvided

async function assertExtensionValuesProvided(data, action) {
  if (action === 'update') {
    await assertHasRequiredProperties(data, ['id'])
  }
  await assertHasRequiredProperties(data.extension, ['name'])
  await assertHasOneProperty(data.extension, ['src', 'srcdoc'])
}

module.exports.assertExtensionValuesProvided = assertExtensionValuesProvided
