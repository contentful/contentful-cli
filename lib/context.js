const { readFile, writeFile } = require('mz/fs')
const { resolve } = require('path')
const { homedir } = require('os')
const findUp = require('find-up')

const { proxyStringToObject } = require('./utils/proxy')

let context
let configPath

async function getConfigPath() {
  const contentfulrc = '.contentfulrc.json'
  const defaultPath = resolve(homedir(), contentfulrc)
  const nestedConfigPath = await findUp(contentfulrc)
  configPath = nestedConfigPath || defaultPath
  return configPath
}

module.exports.getConfigPath = getConfigPath

async function getContext() {
  if (!context) {
    context = await loadRuntimeConfig()
  }

  return context
}

module.exports.getContext = getContext

function setContext(newContext) {
  context = {
    ...context,
    ...newContext
  }
}

module.exports.setContext = setContext

function emptyContext() {
  context = null
}

module.exports.emptyContext = emptyContext

function loadProxyFromEnv(env) {
  const proxyString =
    env['https_proxy'] ||
    env['HTTPS_PROXY'] ||
    env['http_proxy'] ||
    env['HTTP_PROXY']

  // Delete all potential proxy keys `axios` might use
  delete env['http_proxy']
  delete env['HTTP_PROXY']
  delete env['https_proxy']
  delete env['HTTPS_PROXY']

  if (!proxyString) {
    return {}
  }

  const proxy = proxyStringToObject(proxyString)
  return { proxy }
}

async function loadRuntimeConfig() {
  const configPath = await getConfigPath()
  let configFileContent

  try {
    const content = await readFile(configPath)
    configFileContent = JSON.parse(content.toString())
  } catch (e) {
    if (e.code !== 'ENOENT') {
      throw e
    }

    configFileContent = {}
  }

  const environment = loadProxyFromEnv(process.env)

  context = {
    ...context,
    ...configFileContent,
    ...environment
  }

  return context
}

async function storeRuntimeConfig() {
  const contextToStore = {
    managementToken: context.managementToken,
    activeSpaceId: context.activeSpaceId,
    activeEnvironmentId: context.activeEnvironmentId,
    host: context.host,
    proxy: context.proxy,
    rawProxy: context.rawProxy
  }
  return writeFile(
    await getConfigPath(),
    JSON.stringify(contextToStore, null, 2) + '\n'
  )
}

module.exports.storeRuntimeConfig = storeRuntimeConfig
