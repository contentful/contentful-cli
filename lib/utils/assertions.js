const { getContext } = require('../context')
const { PreconditionFailedError } = require('./error')
const { highlightStyle } = require('./styles')

async function assertLoggedIn({ managementToken, paramName } = {}) {
  let noToken = !managementToken
  if (!managementToken) {
    const context = await getContext()
    noToken = !context.managementToken
  }
  paramName = paramName || '--management-token'
  if (noToken) {
    throw new PreconditionFailedError(
      `You have to be logged in to do this.\nYou can log in via ${highlightStyle(
        'contentful login'
      )}\nOr provide a management token via ${paramName} argument`
    )
  }
}

module.exports.assertLoggedIn = assertLoggedIn

async function assertSpaceIdProvided({ spaceId, activeSpaceId } = {}) {
  let noSpaceId = !spaceId && !activeSpaceId
  if (noSpaceId) {
    const context = await getContext()
    noSpaceId = !context.activeSpaceId
  }
  if (noSpaceId) {
    throw new PreconditionFailedError(
      `You need to provide a space id. You can pass it via the ${highlightStyle(
        '--space-id'
      )} parameter or by running ${highlightStyle('contentful space use')}`
    )
  }
}

module.exports.assertSpaceIdProvided = assertSpaceIdProvided
