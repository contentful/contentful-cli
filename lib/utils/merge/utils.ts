import type { Operation } from './types'
import { ContentTypeProps } from 'contentful-management'

export const getChangeKey = (operation: Operation): string => {
  const parts = operation.path.split('/')
  return parts[parts.length - 1]
}

export const getFieldIdForIndex = (
  model: ContentTypeProps[],
  contentTypeId: string,
  index: number
): string => {
  const matchedContentType = model.find(
    contentType => contentType.sys.id === contentTypeId
  )
  if (matchedContentType && matchedContentType.fields.length > index) {
    return matchedContentType.fields[index].id
  }
  return 'unknown'
}

export const fieldIndex = (path: string): number => {
  if (!path.startsWith('/fields')) {
    return -1
  }
  return parseInt(path.split('/')[2])
}

export const fieldChange = (operation: Operation): string => {
  if (!operation.path.startsWith('/fields')) {
    return ''
  }
  return operation.path.split('/')[3]
}

// Util functions for move operations

export const isNestedMoveOperation = (operation: Operation): boolean => {
  return (
    operation.op === 'move' &&
    operation.path.split('/').filter(x => x !== '').length > 2
  )
}

export const getLastIndexFromPath = (path: string): number => {
  const stringIndex = path.split('/').pop() as string
  const lastIndex = parseInt(stringIndex, 10)
  if (isNaN(lastIndex)) {
    return -1
  }
  return lastIndex
}

export const getNestedPropertyNames = (operation: Operation) =>
  operation.path
    .split('/')
    // we want any property that is within fields, we filter out indices
    .filter(x => x !== 'fields' && x !== '' && isNaN(parseInt(x, 10)))
