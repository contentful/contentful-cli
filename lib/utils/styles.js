import chalk from 'chalk'

export const successStyle = chalk.green
export const warningStyle = chalk.yellow
export const errorStyle = chalk.red
export const highlightStyle = chalk.cyan
export const codeStyle = chalk.dim
export const pathStyle = chalk.cyan.underline
