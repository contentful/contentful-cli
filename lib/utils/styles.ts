const chalk = require('chalk')

chalk.enabled = process.env.NODE_ENV !== 'test'

module.exports.successStyle = chalk.green
module.exports.warningStyle = chalk.yellow
module.exports.errorStyle = chalk.red
module.exports.highlightStyle = chalk.cyan
module.exports.headingStyle = chalk.cyan.bold
module.exports.codeStyle = chalk.dim
module.exports.pathStyle = chalk.cyan.underline
module.exports.pathTextStyle = chalk.blue
module.exports.blockQuoteStyle = chalk.magenta.italic
module.exports.htmlStyle = chalk.green
