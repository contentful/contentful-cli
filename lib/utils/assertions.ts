import { getContext } from '../context'
import { PreconditionFailedError } from './error'
import { highlightStyle } from './styles'

interface AssertLoggedInOptions {
  managementToken?: string
  paramName?: string
}

export async function assertLoggedIn({
  managementToken,
  paramName
}: AssertLoggedInOptions = {}): Promise<void> {
  let noToken = !managementToken
  if (!managementToken) {
    const context = await getContext()
    noToken = !context.managementToken
  }
  paramName = paramName || '--management-token'
  if (noToken) {
    throw new PreconditionFailedError(
      `You have to be logged in to do this.\nYou can log in via ${highlightStyle(
        'contentful login'
      )}\nOr provide a management token via ${paramName} argument`
    )
  }
}

interface AssertSpaceIdProvidedOptions {
  spaceId?: string
  activeSpaceId?: string
}

export async function assertSpaceIdProvided({
  spaceId,
  activeSpaceId
}: AssertSpaceIdProvidedOptions = {}): Promise<void> {
  let noSpaceId = !spaceId && !activeSpaceId
  if (noSpaceId) {
    const context = await getContext()
    noSpaceId = !context.activeSpaceId
  }
  if (noSpaceId) {
    throw new PreconditionFailedError(
      `You need to provide a space id. You can pass it via the ${highlightStyle(
        '--space-id'
      )} parameter or by running ${highlightStyle('contentful space use')}`
    )
  }
}
