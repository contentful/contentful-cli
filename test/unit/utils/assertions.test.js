import { assertLoggedIn, assertSpaceIdProvided } from '../../../lib/utils/assertions'

import { getContext } from '../../../lib/context'
import { highlightStyle } from '../../../lib/utils/styles'

jest.mock('../../../lib/context')
jest.mock('../../../lib/utils/styles')

getContext.mockResolvedValue({
  cmaToken: 'mockedToken'
})

afterEach(() => {
  highlightStyle.mockClear()
  getContext.mockClear()
})

test('assertLoggedIn when not logged in', async () => {
  getContext.mockResolvedValueOnce({})
  await expect(assertLoggedIn()).rejects.toThrowErrorMatchingSnapshot()
})

test('assertLoggedIn when logged in', async () => {
  await expect(assertLoggedIn)
})

test('assertSpaceIdProvided when provided via args', async () => {
  await assertSpaceIdProvided({
    spaceId: 'mocked spaceId'
  })
})

test('assertSpaceIdProvided when provided via context', async () => {
  getContext.mockResolvedValueOnce({
    activeSpaceId: 'space id',
    cmaToken: 'mockedToken'
  })
  await assertSpaceIdProvided()
})

test('assertSpaceIdProvided when not provided at all', async () => {
  await expect(assertSpaceIdProvided({})).rejects.toThrowErrorMatchingSnapshot()
})
