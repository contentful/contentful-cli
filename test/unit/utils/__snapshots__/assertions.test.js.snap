// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`assertLoggedIn when not logged in 1`] = `
"You have to be logged in to do this.
You can log in via undefined
Or provide a managementToken via --management-Token argument"
`;

exports[`assertSpaceIdProvided when not provided at all 1`] = `"You need to provide a space id. You can pass it via the undefined parameter or by running undefined"`;
